psc-flink/src/test/java/com/pinterest/flink/streaming/connectors/psc/table/PscDynamicTableFactoryTest.java

```diff
@@ -1343,6 +1343,172 @@ public void testTableSourceWithFractionalRateLimit() {
+        // Since rescale is disabled, there should be no intermediate rescale transformation
+        assertThat(transformation).isNotNull();
+        assertThat(transformation.getName()).contains("PscRateLimit");
+    }
+
+    @Test
+    public void testOperatorChainingWithRescaleAndRateLimit() {
+        // When both rescale and rate limiting are enabled, verify configuration
+        final Map<String, String> modifiedOptions =
+                getModifiedOptions(
+                        getBasicSourceOptions(),
+                        options -> {
+                            options.put("scan.enable-rescale", "true");
+                            options.put("scan.rate-limit.records-per-second", "5000.0");
+                            options.put("scan.parallelism", "100");  // Explicitly set higher than partition count
+                        });
+        
+        final DynamicTableSource actualSource = createTableSource(SCHEMA, modifiedOptions);
+        assertThat(actualSource).isInstanceOf(PscDynamicSource.class);
+        
+        final PscDynamicSource pscSource = (PscDynamicSource) actualSource;
+        
+        // Verify configuration is correctly set
+        assertThat(pscSource.rateLimitRecordsPerSecond).isNotNull();
+        assertThat(pscSource.rateLimitRecordsPerSecond).isEqualTo(5000.0);
+        assertThat(pscSource.scanParallelism).isEqualTo(100);
+        
+        // Get the runtime provider and produce the data stream
+        final ScanTableSource.ScanRuntimeProvider runtimeProvider = 
+                pscSource.getScanRuntimeProvider(ScanRuntimeProviderContext.INSTANCE);
+        assertThat(runtimeProvider).isInstanceOf(DataStreamScanProvider.class);
+        
+        final DataStreamScanProvider dataStreamProvider = (DataStreamScanProvider) runtimeProvider;
+        
+        // Create a local environment with controlled parallelism
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironment();
+        env.setParallelism(10);  // Default parallelism
+        
+        // Produce the data stream and get the transformation chain
+        final Transformation<RowData> transformation = 
+                dataStreamProvider.produceDataStream(n -> Optional.empty(), env).getTransformation();
+        
+        // The final transformation should be the rate limiter
+        assertThat(transformation).isNotNull();
+        assertThat(transformation.getName()).contains("PscRateLimit");
+        
+        // In test environment without real partitions, runtime rescale decision may differ
+        // The key verification is that configuration is properly passed through
+        // In production, with actual partition count > scan.parallelism, rescale would be applied
+    }
+
+    @Test
+    public void testOperatorChainingWithRescaleWithoutRateLimit() {
+        // When rescale is enabled but rate limiting is disabled
```

Comment:
The test prep code are pretty repetitive. Can we fold it into a single util method?